wire-ios-sync-engine/Source/Synchronization/SyncAgent.swift

```diff
@@ -0,0 +1,149 @@
+//
+// Wire
+// Copyright (C) 2025 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireDataModel
+import WireDomain
+import WireLogging
+import WireUtilities
+
+// TODO: [WPB-15440] remove objc interoperability.
+// To temporarily bridge this to legacy code, this inherits from NSObject
+// and exposes a method to objc. Once we integrate the new incremental
+// sync, we won't need to bridge to legacy code and remove the inheritance.
+
+final class SyncAgent: NSObject {
+
+    weak var delegate: SyncAgentDelegate?
+    private let lastUpdateEventIDRepository: any LastEventIDRepositoryInterface
+    private let initialSyncBuilder: any InitialSyncBuilderProtocol
+    private let legacySyncStatus: any SyncStatusProtocol
+
+    private var hasPerformedInitialSync: Bool {
+        lastUpdateEventIDRepository.fetchLastEventID() != nil
+    }
+
+    // MARK: - Life cycle
+
+    init(
+        lastUpdateEventIDRepository: any LastEventIDRepositoryInterface,
+        initialSyncBuilder: any InitialSyncBuilderProtocol,
+        legacySyncStatus: any SyncStatusProtocol
+    ) {
+        self.lastUpdateEventIDRepository = lastUpdateEventIDRepository
+        self.initialSyncBuilder = initialSyncBuilder
+        self.legacySyncStatus = legacySyncStatus
+        super.init()
+    }
+
+    // MARK: - API
+
+    /// Performs the appropriate sync depending in the local state.
+    ///
+    /// If no last event id is known, then the initial sync will be performed,
+    /// otherwise the incremental sync will be performed.
+
+    func performSyncIfNeeded() async throws {
+        if !hasPerformedInitialSync {
+            try await performInitialSync()
+        } else {
+            try await performIncrementalSync()
+        }
```

Comment:
Ideally we would perform initial sync, then incremental sync, one after the other. But the legacy initial sync continues automatically into the incremental sync. For this reason, I've also made the initial sync behave like this... when it finishes we start automatically the incremental sync. Thus, `performingInitialSync` will also invoke `performIncrementalSync`.

Reply:
ok i see this in performingInitialSync